src/live_restore/live_restore_state.c

```diff
@@ -0,0 +1,362 @@
+    else
+        WT_RET_MSG(session, EINVAL, "Invalid state string: '%s' ", state_str);
+
+    return (0);
+}
+
+/*
+ * __live_restore_get_state_from_file --
+ *     Read the live restore state from the turtle file. If it doesn't exist we return NONE. The
+ *     caller must already hold the live restore state lock. This function takes a non-live restore
+ *     file system as it can be called in non-live restore modes.
+ */
+static int
+__live_restore_get_state_from_file(
+  WT_SESSION_IMPL *session, WT_FILE_SYSTEM *fs, WTI_LIVE_RESTORE_STATE *statep)
+{
+    WT_DECL_RET;
+
+    bool turtle_in_dest = false;
+    WTI_LIVE_RESTORE_STATE state_from_file = WTI_LIVE_RESTORE_STATE_NONE;
+    char *lr_metadata = NULL;
+
+    WT_DECL_ITEM(dest_turt_path);
+    WT_RET(__wt_scr_alloc(session, 0, &dest_turt_path));
+    WT_ERR(__wt_filename_construct(
+      session, S2C(session)->home, WT_METADATA_TURTLE, UINTMAX_MAX, UINT32_MAX, dest_turt_path));
+
+    WT_ERR(
+      fs->fs_exist(fs, (WT_SESSION *)session, (char *)(dest_turt_path)->data, &turtle_in_dest));
+
+    if (!turtle_in_dest)
+        state_from_file = WTI_LIVE_RESTORE_STATE_NONE;
+    else {
+        WT_ERR_NOTFOUND_OK(
+          __wt_metadata_search(session, WT_METADATA_LIVE_RESTORE, &lr_metadata), true);
+        if (ret == WT_NOTFOUND) {
+            state_from_file = WTI_LIVE_RESTORE_STATE_NONE;
+            ret = 0;
+        } else {
+            char lr_metadata_string[WT_LIVE_RESTORE_STATE_STRING_MAX];
+            if ((sscanf(lr_metadata, "state=%127s", lr_metadata_string)) != 1)
+                WT_ASSERT_ALWAYS(
+                  session, false, "failed to parse live restore metadata from the turtle file!");
+
+            __live_restore_state_from_string(session, lr_metadata_string, &state_from_file);
+        }
+    }
+
+    *statep = state_from_file;
+
+err:
+
+    __wt_free(session, lr_metadata);
+    __wt_scr_free(session, &dest_turt_path);
+
+    return (ret);
+}
+
+/*
+ * __live_restore_report_state_to_application --
+ *     WiredTiger reports a simplified live restore state to the application which lets it know it
+ *     can restart on completion of live restore.
+ */
+static void
+__live_restore_report_state_to_application(WT_SESSION_IMPL *session, WTI_LIVE_RESTORE_STATE state)
+{
+    switch (state) {
+    case WTI_LIVE_RESTORE_STATE_NONE:
+        WT_STAT_CONN_SET(session, live_restore_state, WT_LIVE_RESTORE_INIT);
+        break;
+    case WTI_LIVE_RESTORE_STATE_BACKGROUND_MIGRATION:
+    case WTI_LIVE_RESTORE_STATE_CLEAN_UP:
+        WT_STAT_CONN_SET(session, live_restore_state, WT_LIVE_RESTORE_IN_PROGRESS);
+        break;
+    case WTI_LIVE_RESTORE_STATE_COMPLETE:
+        WT_STAT_CONN_SET(session, live_restore_state, WT_LIVE_RESTORE_COMPLETE);
+        break;
+    }
+}
+
+/*
+ * __wti_live_restore_set_state --
+ *     Update the live restore state in memory and persist it to the turtle file.
+ */
+int
+__wti_live_restore_set_state(
```

Comment:
A more traditional usage of spinlocks would expect callers of this to use: WT_WITH_LOCK_WAIT, that handles re-entrantcy. I don't think we should do that cleanup in this PR but I think it may be worth a FIXME to investigate.